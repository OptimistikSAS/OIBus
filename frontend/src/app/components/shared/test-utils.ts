import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
