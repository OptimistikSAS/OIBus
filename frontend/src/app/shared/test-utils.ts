import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
