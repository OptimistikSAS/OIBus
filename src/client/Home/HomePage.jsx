import React from 'react'
import { Spinner } from 'reactstrap'
import apis from '../services/apis'
import { AlertContext } from '../context/AlertContext.jsx'
import { ConfigContext } from '../context/configContext.jsx'
import NodeView from './NodeView.jsx'
import utils from '../helpers/utils'

const HomePage = () => {
  const [loading, setLoading] = React.useState(false)
  const [status, setStatus] = React.useState({})
  const { setAlert } = React.useContext(AlertContext)
  const { activeConfig } = React.useContext(ConfigContext)
  const config = utils.jsonCopy(activeConfig)
  utils.replaceValues(config, ['password', 'secretKey'], '******')

  /**
   * Acquire the status
   * @returns {void}
   */
  const fetchStatus = () => {
    apis
      .getStatus()
      .then((response) => {
        setStatus(response)
      })
      .catch((error) => {
        console.error(error)
        setAlert({ text: error.message, type: 'danger' })
      })
  }

  /**
   * Fetch status after render
   * @returns {void}
   */
  React.useEffect(() => {
    fetchStatus()
  }, [])

  // Disable loading when server reachable
  const stopLoadingWhenReachable = () => {
    apis
      .getConfig()
      .then(() => setLoading(false))
      // retry getConfig if error catched
      .catch(() => setTimeout(() => stopLoadingWhenReachable(), 1000))
  }

  /**
   * Restart request
   * @returns {void}
   */
  const handleRestart = async () => {
    try {
      await apis.reload()
      setLoading(true)
      // start checking if server is reachable after 10 sec,
      // as the restart on backend has a 10 sec delay
      setTimeout(() => stopLoadingWhenReachable(), 10000)
    } catch (error) {
      console.error(error)
      setAlert({ text: error.message, type: 'danger' })
    }
  }

  /**
   * Shutdown request
   * @returns {void}
   */
  const handleShutdown = async () => {
    try {
      await apis.shutdown()
    } catch (error) {
      console.error(error)
      setAlert({ text: error.message, type: 'danger' })
    }
  }

  return (
    <>
      <NodeView status={status} onRestart={handleRestart} onShutdown={handleShutdown} />
      {loading && (
        <div className="spinner-container">
          <Spinner color="primary" />
        </div>
      )}
    </>
  )
}
export default HomePage
