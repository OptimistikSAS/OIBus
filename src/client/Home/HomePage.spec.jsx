/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'
import HomePage from './HomePage.jsx'

import activeConfig from '../../../tests/testConfig'

// ReacFlow does not seem to be working with jest.
// so we have to mock this component
jest.mock('../../../node_modules/react-flow-renderer/dist/ReactFlow.js', () => () => 'ReactFlow')

global.EventSource = class {
  constructor() {
    this.close = () => {}
  }
}

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

const setAlert = jest.fn()

React.useContext = jest.fn().mockReturnValue({ activeConfig, setAlert })
describe('HomePage', () => {
  test('display Health page based on config', async () => {
    act(() => {
      ReactDOM.render(
        <HomePage />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })
  test('display Health with config null', () => {
    React.useContext = jest.fn().mockReturnValue({ activeConfig: null, setAlert })
    act(() => {
      ReactDOM.render(
        <HomePage />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
    React.useContext = jest.fn().mockReturnValue({ activeConfig, setAlert })
  })
})
