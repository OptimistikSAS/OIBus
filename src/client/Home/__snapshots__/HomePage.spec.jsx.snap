// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage display Health page based on config 1`] = `
<div>
  <div
    class="container"
  >
    <div
      style="height: 710px; width: 100%; padding-top: 2rem;"
    >
      ReactFlow
    </div>
  </div>
</div>
`;

exports[`HomePage display Health with config null 1`] = `
<div>
  <div
    class="container"
  >
    <div
      style="height: 710px; width: 100%; padding-top: 2rem;"
    >
      ReactFlow
    </div>
  </div>
</div>
`;
